ape_hardhat/__init__.py

```diff
@@ -0,0 +1,237 @@
+"""
+Ape network provider plugin for Hardhat (Ethereum development framework and network
+implementation written in Node.js).
+"""
+
+import atexit
+import ctypes
+import platform
+import random
+import signal
+import sys
+import time
+from subprocess import PIPE, Popen, call
+from typing import Any, List, Optional
+
+from ape import plugins
+from ape_http.providers import DEFAULT_SETTINGS, EthereumProvider, NetworkConfig
+
+EPHEMERAL_PORTS_START = 49152
+EPHEMERAL_PORTS_END = 60999
+HARDHAT_CHAIN_ID = 31337
+HARDHAT_CONFIG = """
+// See https://hardhat.org/config/ for config options.
+module.exports = {
+};
+"""
+HARDHAT_START_NETWORK_RETRIES = [0.1, 0.2, 0.3, 0.5, 1.0]  # seconds between network retries
+HARDHAT_START_PROCESS_ATTEMPTS = 3  # number of attempts to start subprocess before giving up
+
+
+def _signal_handler(signum, frame):
+    """Runs on SIGTERM and SIGINT to force ``atexit`` handlers to run."""
+    atexit._run_exitfuncs()
+    sys.exit(143 if signum == signal.SIGTERM else 130)
+
+
+def _set_death_signal():
+    """Automatically sends SIGTERM to child subprocesses when parent process dies."""
+    if platform.uname().system == "Linux":
+        # from: https://stackoverflow.com/a/43152455/75956
+        # the first argument "1" is PR_SET_PDEATHSIG
+        # the second argument is what signal to send the child
+        libc = ctypes.CDLL("libc.so.6")
+        return libc.prctl(1, signal.SIGTERM)
+
+
+class HardhatSubprocessError(RuntimeError):
+    pass
+
+
+class HardhatNetworkConfig(NetworkConfig):
+    # --fork <URL, JSON-RPC server to fork from>
+    fork_url: Optional[str] = None
+
+    # --fork-block-number <INT, block number to fork from>
+    fork_block_number: Optional[int] = None
```

Comment:
I think forking stuff should be left out of the config. I wanted to explore a more comprehensive "fork mode support" at a later date, as it's sort of a modifier to the provider during live operation vs. a separate network type. something like:
```python
with provider.temporary_fork(...):
    ...
```

Reply:
A context manager like that would be handy. Right now I'm just exposing Hardhat's underlying ability to fork via these configs. Do you want me to remove the ability to fork for now? I think some people might find that feature useful even without the nicer context manager. But it's easy to remove. LMK and I'll do that before landing.

Reply:
take these config options out for now, don't want to ingrain a bad habit here